scripts/compaction_bounds.zig

```diff
@@ -0,0 +1,128 @@
+                        credits_pending: u64,
+                        credits_posted: u64,
+                    }
+                else
+                    // Secondary index.
+                    IndexCompositeKeyType(@TypeOf(@field(@as(Transfer, undefined), field_name)));
+
+                const values_per_table = @divTrunc(config.lsm_table_size_max, @sizeOf(Value));
+                const table_count = try std.math.divCeil(usize, existing_transfer_count, values_per_table);
+
+                var level_count: usize = 0;
+                {
+                    var table_count_remaining: usize = table_count;
+                    var tables_per_level: usize = config.lsm_growth_factor;
+                    while (table_count_remaining > 0) {
+                        level_count += 1;
+                        table_count_remaining = table_count_remaining -| tables_per_level;
+                        tables_per_level *= config.lsm_growth_factor;
+                    }
+                }
+
+                // Write bandwidth for flushing the immutable table:
+                bandwidth_bytes_per_second +=
+                    // Total size of all values
+                    transfers_per_second * @sizeOf(Value) *
+                    // All trees except the object tree might have to write tombstones
+                    @as(usize, if (std.meta.eql(field_name, "timestamp")) 1 else 2);
+
+                // Read/write bandwidth for compaction:
+                const bandwidth_bytes_per_compaction =
+                    // Compaction touches:
+                    // * One table from level a.
+                    // * Up to config.lsm_growth_factor tables from level b.
+                    // TODO Compaction for immutable table to level 0 doesn't have to read immutable table from disk.
+                    (1 + config.lsm_growth_factor) *
+                    // The data in each table must be read from disk and written back out.
+                    // TODO This factor could actually be as high as 4 with tombstones.
+                    2 * config.lsm_table_size_max;
+                // The object tree does not need to be compacted.
+                if (!std.meta.eql(field_name, "timestamp")) {
+                    bandwidth_bytes_per_second +=
+                        // In the ideal case we flush the immutable table when it contains exactly `values_per_table` transfers.
+                        (try std.math.divCeil(usize, transfers_per_second, values_per_table)) *
+                        // For every table ingested at level 0, we have to compact into each level once.
+                        level_count *
+                        bandwidth_bytes_per_compaction;
+                    switch (case) {
+                        .ideal => {},
+                        .actual => bandwidth_bytes_per_second +=
+                            // Currently, we flush every `config.lsm_batch_multiple` ops, causing many extra level 0 compactions.
```

Comment:
Are we overstating this given that secondary tables in level 0 are typically a function of `commit_entries_max` not `table_size_max`? i.e. On the order of 10x smaller?

Reply:
Do you mean the immutable table? I didn't think that the size of the tables in the lsm changed depending on the tree. That would change the calculation a lot.

Reply:
I've changed the calculation to correct for the size of the immutable table. Does it look correct now?

Reply:
Yes, I mean the immutable and mutable tables.

The size of these tables in the LSM do in fact depend on the tree, since the mutable/immutable tables are sized not by `table_size_max` (an upper bound) but rather by `commit_entries_max * value_size` and `value_size` changes according to tree.

The size of the mutable/immutable tables is what determines ingress rate, and therefore egress rate across all levels, since our invariant is that compaction is only required to compact as little as needed across levels to flush the next immutable table.

Reply:
It sounds like you are saying that the number of compactions per second in each of the lower levels is `transfers_per_second / commit_entries_max` rather than `transfers_per_second / values_per_table`. If that is the case then we will do substantially _more_ compactions than is simulated above.

Reply:
No, I don't mean `transfers_per_second / commit_entries_max`.

Rather that we should only be compacting `commit_entries_max` instead of `values_per_table` per compaction. How `commit_entries_max` relates to `transfers_per_second` depends on the tree, it can't be related directly.